Add "transforming telemetry" section to collector docs (#2304)
open-telemetry/opentelemetry.io: content/en/docs/collector/transforming-telemetry.md

Processors available from the the [Collector Contrib repository](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor) support dozens of different transformations on metric, span and log data. The following sections provide some basic examples on getting started with a few frequently-used processors.

The configuration of processors, particularly advanced transformations, may have a signficant impact on collector performance.

## Basic filtering

**Processor**: [filter processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/filterprocessor)

The filter processor allows users to filter telemetry based on `include` or `exclude` rules. Include rules are used for defining "allow lists" where anything that does _not_ match include rules is dropped from the collector. Exclude rules are used for defining 
"deny lists" where telemetry that matches rules is dropped from the collector.

For example, to _only_ allow span data from services app1, app2, and app3 and drop data from all other services:

> ```

Reviewer: Add a code-block language as appropriate, here and elsewhere in the document:

```suggestion
```yaml
```

Reviewer: > Add a code-block language as appropriate, here and **_elsewhere_**

You'll also need to add languages to the other code blocks. Marking this as unresolved until then. Thanks.

Reviewer 2: @smithclay please resolve the conversation if this has been addressed

processors:
  filter/allowlist:
    spans:
      include:
        match_type: strict
        services:
          - app1
          - app2
          - app3
```

To only block spans from a service called development while allowing all other spans, an exclude rule is used:

```
processors:
  filter/denylist:
    spans:
      exclude:
        match_type: strict
        services:
          - development
```

> The [filter processor docs](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/filterprocessor) have more examples, including filtering on logs and metrics. More advanced filtering is also available in the [transform processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor).

Reviewer 2: You mention twice that the transform processor allows advanced things, maybe having a final section "Advanced Transformations" that calls out the existence of the transform processor and especially the transformation query language?

Reviewer 2: @smithclay please resolve the conversation if this has been addressed

```
processors:
  resource/k8s:
    attributes:
    - key: k8s.cluster.name
      from_attribute: k8s-cluster
      action: insert
```

More advanced attribute transformations are also available in the [transform processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor).

## Renaming Metrics or Metric Labels

**Processor:** [metrics transform processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/metricstransformprocessor)

The [metrics transform processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/metricstransformprocessor) shares some functionality with the [attributes processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/attributesprocessor), but also supports renaming and other metric-specific functionality.

```
processors:
  metricstransform/rename:
    transforms:
      include: system.cpu.usage`
      action: update
      new_name: system.cpu.usage_time
```

> The [metrics transform processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/metricstransformprocessor)` `also supports regexes to apply transform rules to multiple metric names or metric labels at the same time. This example renames cluster_name to cluster-name for all metrics:

Reviewer 2: ```suggestion
The [metrics transform processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/metricstransformprocessor) also supports regexes to apply transform rules to multiple metric names or metric labels at the same time. This example renames cluster_name to cluster-name for all metrics:
```

```
processors:
  metricstransform/clustername:
   transforms:
     - include: "^.*$"
       match_type: regexp
       action: update
       operations:
       - action: update_label
         label: cluster_name
         new_label: cluster-name
```

## Enriching Telemetry with Resource Attributes

**Processor**: [resource detection processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/resourcedetectionprocessor) and [k8sattributes processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/k8sattributesprocessor)

> It is recommended that end-users use processors to enrich their telemetry data with relevant infrastructure metadata to help teams quickly identify when underlying infrastructure is impacting service health or performance.

Reviewer 3: I'd remove the recommendation part and consider for example:
"It is recommended that end-users use processors to enrich their telemetry data with..."
->
"These processors can be used for enriching telemetry with..."
(this is to keep the style similar to other parts in this document)

> Lightstep recommends that customers use processors to enrich their telemetry data with relevant infrastructure metadata to help teams quickly identify when underlying infrastructure is impacting service health or performance.

Reviewer 2: Please remove vendor reference here
```suggestion
It is recommended that end-users use processors to enrich their telemetry data with relevant infrastructure metadata to help teams quickly identify when underlying infrastructure is impacting service health or performance.
```

---
title: Transforming telemetry

> weight: 31

Reviewer: Change the weight so that it appears below Scaling, and before all the "Build" pages. (Right now it's in the middle of the "Build" pages -- see below.)

```suggestion
weight: 26
```

We'll have to space those weights out in a followup PR. /cc @svrnm @cartermp 

> <img width="285" alt="image" src="https://user-images.githubusercontent.com/4140793/218588654-e3f65271-8981-4510-8bfd-9228c920d602.png">
